File: design-underground-system/design-underground-system.go
type UndergroundSystem struct {
    timeMap map[string]int
    countMap map[string]int
    tripLocMap map[int]string
    tripTimeMap map[int]int
}
​
​
func Constructor() UndergroundSystem {
    timeMap := make(map[string]int)
    countMap := make(map[string]int)
    tripLocMap := make(map[int]string)
    tripTimeMap := make(map[int]int)
    return UndergroundSystem{timeMap, countMap, tripLocMap, tripTimeMap}
}
​
​
func (this *UndergroundSystem) CheckIn(id int, stationName string, t int)  {
    this.tripLocMap[id] = stationName
    this.tripTimeMap[id] = t
}
​
​
func (this *UndergroundSystem) CheckOut(id int, stationName string, t int)  {
    loc, _ := this.tripLocMap[id]
    time, _ := this.tripTimeMap[id]
    
    delete(this.tripLocMap, id)
    delete(this.tripTimeMap, id)
    
    key := loc + stationName
    this.timeMap[key] += t - time
    this.countMap[key] += 1
}
​
​
func (this *UndergroundSystem) GetAverageTime(startStation string, endStation string) float64 {
    key := startStation + endStation
    time, _ := this.timeMap[key]
    count, _ := this.countMap[key]
    return float64(time) / float64(count)
}
​
​
/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * obj := Constructor();
 * obj.CheckIn(id,stationName,t);
 * obj.CheckOut(id,stationName,t);
 * param_3 := obj.GetAverageTime(startStation,endStation);
 */

File: design-underground-system/design-underground-system.java
class UndergroundSystem {
    Map<Integer,Integer> checkIn;
    Map<Integer,String> src;
    Map<String,int[]> times;
    public UndergroundSystem() {
        checkIn = new HashMap<>();
        times = new HashMap<>();
        src = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        checkIn.put(id, t);
        src.put(id, stationName);
    }
    
    public void checkOut(int id, String stationName, int t) {
        int timeIn = checkIn.get(id);
        String loc = src.get(id);
        String strId = loc + stationName;
        int[] existing = times.getOrDefault(strId, new int[2]);
        existing[0] += 1;
        existing[1] += t - timeIn;
        times.put(strId, existing);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        int[] time = times.get(startStation + endStation);
        return (double) time[1] / time[0];
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */